spring-xd-dirt/src/main/java/org/springframework/xd/dirt/module/ResourceModuleRegistry.java

```
@@ -16,25 +16,319 @@
- * <i>e.g.</i> {@code source/time/lib/*.jar}</li>
- * </ul>
- * 
- * @author Mark Fisher
- * @author Glenn Renfro
+ * {@link Resource} based implementation of {@link ModuleRegistry} that supports mutative operations as well.
+ *
+ * <p>Will generate MD5 hash files for written modules. Also, while not being coupled to {@code java.io.File} access,
+ * will try to sanitize file paths as much as possible.</p>
+ *
  * @author Eric Bottard
  */
-public class ResourceModuleRegistry extends ArchiveModuleRegistry {
+public class ResourceModuleRegistry implements WritableModuleRegistry, InitializingBean {
+
+	/**
+	 * The extension the module 'File' must have if it's not in exploded dir format.
+	 */
+	public static final String ARCHIVE_AS_FILE_EXTENSION = ".jar";
+
+	/**
+	 * The extension to use for storing an uploaded module hash.
+	 */
+	public static final String HASH_EXTENSION = ".md5";
+
+	private final static String[] SUFFIXES = new String[] {"", ARCHIVE_AS_FILE_EXTENSION};
+
+	final protected static byte[] HEX_DIGITS = "0123456789ABCDEF".getBytes();
+
+	private boolean allowWrite;
+
+	/**
+	 * Whether to require the presence of a hash file for archives. Helps preventing half-uploaded archives from
+	 * being picked up.
+	 */
+	private boolean requireHashFiles;
+
+	/**
+	 * Whether to attempt to create the directory structure at startup (disable for read-only implementations).
+	 */
+	private boolean createDirectoryStructure;
+
+	private ResourcePatternResolver resolver = new PathMatchingResourcePatternResolver();
+
+	private String root;
 
 	public ResourceModuleRegistry(String root) {
-		super(root);
+		this(root, false);
+	}
+
+	public ResourceModuleRegistry(String root, boolean allowWrite) {
+		this.allowWrite = this.createDirectoryStructure = this.requireHashFiles = allowWrite;
+		this.root = StringUtils.trimTrailingCharacter(root, '/');
+	}
+
+	@Override
+	public boolean delete(ModuleDefinition definition) {
+		if (!allowWrite) {
```

It seems like this should throw an UnsupportedOperationException